### DISCUSSION TITLE
Other MiniDSO devices

### DISCUSSION THREAD

**Discussion by nieldk** [gabonator/LA104#32] [Ideas]
How difficult would it be too add support for other DSO devices, like i dont have DS203, but I have Ds202.
I will have a go at it, but if there would be too many difference, for example in the hardware, would you think its doable, or would reversing everything be needed

**Comment by gabonator** [comment]
Not a problem at all. I guess the older DSO's do not have any FPGA, they just rely on internal ADC, so it should be fairly simple. I put together few thoughts about porting onto other platforms here: https://github.com/gabonator/DS203/issues/47

Just clone this folder: 
https://github.com/gabonator/LA104/tree/master/system/os_host/source/bios/clean

name it "ds202" and implement the "imports.c" file. You will need the source code of original ds202 for doing this, but I think it is open sourced as their other oscilloscopes. There are 6x pixel drawing functions, one function which returns pressed buttons, two functions for reading and writing flash and hardware initialization function.

**Comment by nieldk** [comment]
OK, the source is available. but, i cant locate all these functions. I did find some, but, the definitions differ from that imports.c structure.

here is what I found (added to the imports.c in the folder you mentioned).

Source code for the DS202 I got from here: [http://www.minidso.com/forum.php?mod=viewthread&tid=696&extra=page%3D1](url)

```#include "BIOS.h"

void SetPixel(u16 Color)
{
  asm("MOVS    R2, #0x0800     ");
  asm("MOVT    R2, #0x4800     "); // GPIO Port C Base 0x48000800
  asm("MOVS    R3, #0x0800     "); // GPIO_Pin_11
  asm("STRH    R0, [R2, #0x814]"); // GPIO Port E ODR  0x48001014  
  asm("STRH    R3, [R2, #0x028]"); // GPIO Port C BRR  0x48000828
  asm("NOP");
  asm("STRH    R3, [R2, #0x018]"); // GPIO Port C BSRR 0x48000818
  asm("BX      LR");
}

void ReadStart(void)
{
}

void ReadFinish(void)
{
}

u16 ReadPixel(void )
{
  u32 Tmp;

  LCD_CMD_W(0x2E);
  LCD_PORT_IN();
  LCD_nRD_LOW();
  asm("NOP");
  asm("NOP"); 
  LCD_nRD_HIGH();
  asm("NOP");
  asm("NOP"); 
  LCD_nRD_LOW();
  asm("NOP");
  asm("NOP");
  asm("NOP");
  asm("NOP"); 
  Tmp = READ_LCD();
  LCD_nRD_HIGH();
  Tmp = (Tmp & 0xF800)+((Tmp << 3)& 0x07E0);
  LCD_nRD_LOW();
  asm("NOP"); 
  asm("NOP"); 
  asm("NOP"); 
  asm("NOP"); 
  Tmp |= READ_LCD() >> 11;
  LCD_nRD_HIGH();
  LCD_PORT_OUT();
  return Tmp;
}

void Set_Block(u16 x1, u16 y1, u16 x2, u16 y2)
{
  LCD_CMD_W(0x2A);
  LCD_DAT_W(y1 >> 8);
  LCD_DAT_W(y1 & 255);
  LCD_DAT_W(y2 >> 8);
  LCD_DAT_W(y2 & 255);

  LCD_CMD_W(0x2B);
  LCD_DAT_W(x1 >> 8);
  LCD_DAT_W(x1 & 255);
  LCD_DAT_W(x2 >> 8);
  LCD_DAT_W(x2 & 255);

  LCD_CMD_W(0x2C);           // Ready to write GRAM
}

void xBeep(bool b)
{
}

void Set_Posi(uint_fast16_t x, uint_fast16_t y)
{
}

void EnableUsb(bool enable)
{
}

void HardwareInit()
{
}

uint32_t GetKeys()
{
  return 0;
}

void ExtFlashSecWr(u8* pBuf, u32 WrAddr)
{
  u16 SecSize = __Info(SECTOR);
  u32 i, Addr = WrAddr &  (~(SecSize-1));
  ExtFlashSectorErase(Addr); 
  for(i=0; i<SecSize; i+=PAGE_SIZE){
    ExtFlashPageProg(&pBuf[i], Addr+i, PP);
  }
}

void ExtFlashDataRd(u8* pBuf, u32 RdAddr, u16 Lenght)
{
//  NVIC_DisableIRQ(USB_LP_IRQn);
  ExtFlashWaitForWrEnd();
  ExtFlash_CS_LOW();
  ExtFlashSendByte(READ);
  ExtFlashSendByte((RdAddr & 0xFF0000) >> 16);
  ExtFlashSendByte((RdAddr & 0xFF00) >> 8);
  ExtFlashSendByte( RdAddr & 0xFF);
  for(u16 i=0; i<Lenght; i++) pBuf[i] = ~ExtFlashReadByte();
  ExtFlash_CS_HIGH();
//  NVIC_EnableIRQ(USB_LP_IRQn);
}


uintptr_t GetAttribute(enum EAttribute attribute)
{
  switch (attribute)
  {
    case DeviceType: return (uintptr_t)"CLEAN";
    case VersionDfu: return (uintptr_t)nullptr;
    case VersionHardware: return (uintptr_t)(char*)nullptr;
    case VersionSystem: return (uintptr_t)(char*)nullptr;
    case VersionFpga: return (uintptr_t)(char*)nullptr;
    case SerialNumber: return (uintptr_t)(uint32_t)0;
    case LicenseNumber: return (uintptr_t)0;
    case LicenseValid: return (uintptr_t)0;
    case DisplayType: return (uintptr_t)0;
    case DiskType: return (uintptr_t)0;
    default: return 0;
  }
}
```

**Comment by gabonator** [comment]
Looks good, but you will need also **Set_Posi** function. **ReadXXXX** functions should not be necessary for basic applications. But functions ExtFlashSecWr/ExtFlashDataRd should be very similar, in your case it looks suspicious.

Then you will need to fix these definitions:
  - in os_host/source/bios/Bios.h 
    - LCD::Width, LCD::Height
    - FAT::SectorSize, FAT::SectorCount

  - in os_host/source/bios/fat.h
    - BIOS_FAT_SectorSize, BIOS_FAT_SectorCount

And for simplicity I would not override system USB functions - LA104 and DS213 use custom USB stack and DS203 keeps the original BIOS to handle USB functionality. So just find every occurence of "#if defined(DS203)" or DS203HD and add there DS202 as well. 

DS203HD was attempt to add support for 8MB DS203, but I could not finish it since I do not own this device.

**Comment by gabonator** [comment]
And looking at the source code of orig fs, I would place this into HardwareInit:

```
  __Ctrl(SYS_CFG, RCC_DEV | TIM_DEV | GPIO_OPA | ADC_DAC  );
  SPI_Config();
  SysTick_Config(SystemCoreClock/1000);                 //SysTick = 1mS
  __Ctrl(B_LIGHT, 50);                                  //
  Ctrl(PM_CTRL, CHG_ACT);                               //
  __Ctrl(BUZZVOL, 50);                                  //
  Beep(200);                                            //200mS                                     
  Init_Fat_Value();
  TIM6_Config(); 
  __Ctrl(SMPL_ST, DISABLE);
  __Ctrl(SMPL_ST, SIMULTANEO);
  __Ctrl(SMPLBUF, (u32)Smpl);
  __Ctrl(SMPLNUM, DEPTH[PopMenu1_Value[WIN_Depth]]);
  __Ctrl(SMPL_ST, ENABLE);
```

and add there also
```
  NVIC_SetVectorTable(NVIC_VectTab_FLASH, (uint32_t)g_pfnVectors - NVIC_VectTab_FLASH);
```

... they use USB implementation at the application level, so you will need to base your code on LA104 implementation, not on DS203 as I stated before.

SetPosi:
```
void SetPosi(u16 x0, u16 y0)
{
  LCD_CMD_W(0x2A);
  LCD_DAT_W(0x00);
  LCD_DAT_W(y0 & 255);
  LCD_DAT_W(0x00);
  LCD_DAT_W(239);

  LCD_CMD_W(0x2B);
  LCD_DAT_W(x0 >> 8);
  LCD_DAT_W(x0 & 255);
  LCD_DAT_W(0x01);
  LCD_DAT_W(0x3F);

  LCD_CMD_W(0x2C);
}
```

and I take back what I said about the flash functions. I see that you have just copied the code from orig fw, so it should work

sector size, sector count:
    #define SEC_SIZE          4096
    #define SEC_AMNT          2048

**Reply by nieldk** [reply]
Cool, thank you so much.
Hope DS202 will be a new supported device soon

**Comment by nieldk** [comment]
Hmm this will take some time.

After adding BIOS.h from the original source, it gave a lot of conflicting defintions from the imports.c, I removed the conflicting procedures, and came further.
I copied build_la104.sh to build_ds202.sh, but it contains a lot of references to  STM32F10x, whereas the DS202 uses STM32F30x.
I then changed all these to reference STM32F10x includes from the original source.
This provided more conflicts, and there are still some places in the shared code that refers to STM32F10x.

Still working on it, so we will see :)

**Comment by gabonator** [comment]
Ouch, that is quite serious complication if the device is based on different CPU. You will need to add platform support files, see os_host/library/CMSIS and os_host/library/STM_xxxx_Driver

And then the last thing is the linker script - you need to specify, where in the RAM and FLASH the OS should be placed (see os_host/app.ld):
```
/* Define memory regions. */
MEMORY
{
  rom (rx) : ORIGIN = 0x08008000, LENGTH = 160K
  ram (rwx) : ORIGIN = 0x20000000, LENGTH = 40K
}

_estack = ORIGIN(ram)+LENGTH(ram)-0x100;

```

**Reply by nieldk** [reply]
Yeah, I figured so :)
The os_host libraries in CMSIS, STM32F10x_StdPeriph_Driver and  STM32_USB-FS-Device_Driver needs changes.
I included the CMSIS,  STM32F30x_StdPeriph_Driver and  STM32_USB-FS-Device_Driver from the original sources, but CMSIS and STM32_USB-FS-Device_Driver are used for the already supported devices, so these to will need some merging, not to conflict with the existing codebase thus breaking what is already working ....

**Reply by nieldk** [reply]
But, found out, reading the source. If you put an image on the disk, named 'about.bmp'
BMPv3, 4px dept, 16 color, 250x200
The 'about' menuitem in menu 3, will show that 😁
![IMG_20201214_123339](https://user-images.githubusercontent.com/4446241/102094483-abf98a00-3e22-11eb-847c-528fc777dcbe.jpg)

**Comment by nieldk** [comment]
Kinda stuck on this :(
See the log from build in m repository for this
https://github.com/nieldk/DS202/blob/master/system/os_host/log.txt

**Reply by gabonator** [reply]
implementation of LCD_CMD_W is in LCD.h:
```
#define LCD_CMD_W(Cmd)    LCD_RS_LOW();\
                          LCD_DAT_W(Cmd);\
                          LCD_RS_HIGH()
```
just copy missing functions into imports.c, or copy whole LCD.h into the platform support folder

**Comment by nieldk** [comment]
Great, added a bunch of imports to imports.c, and now I got further (see log.txt)
I had to change function "uintptr_t GetAttribute(uintptr_t attribute)" in imports.c to get to this stage as it throwed an error

``In file included from ../source/interrupt.c:7:
../source/bios/ds202/imports.c:147:29: error: 'enum EAttribute' declared inside parameter list will not be visible outside of this definition or declaration [-Werror]
  147 | uintptr_t GetAttribute(enum EAttribute attribute)'``
Also the case statement in that returned undefined types for DeviceType etc.

But, progress
